// vuestic-ui/dist/web-components/src/components/va-timeline/VaTimelineItem/VaTimelineItem.js
import { defineComponent, h } from "vue";
import { u as useComponentPresetProp } from "../../../composables/useComponentPreset.js";
import { e as extractComponentProps } from "../../../utils/component-options/extract-component-options.js";
import { f as filterComponentProps } from "../../../utils/component-options/filter-props.js";
import { V as VaTimelineSeparator } from "../VaTimelineSeparator/index.js";
import { _ as _export_sfc } from "../../../../plugin-vue_export-helper.js";
var _style_0 = ":root,\n:host {\n  --va-timeline-item-display: flex;\n  --va-timeline-item-flex-direction: column;\n  --va-timeline-item-vertical-flex-direction: row;\n  --va-timeline-item-vertical-flex-wrap: nowrap;\n  --va-timeline-item-vertical-align-items: stretch;\n  /* Title */\n  --va-timeline-item-title-text-align: center;\n  --va-timeline-item-title-color: var(--va-primary);\n  --va-timeline-item-title-font-weight: 700;\n  --va-timeline-item-title-font-size: 0.8rem;\n  --va-timeline-item-title-text-transform: uppercase;\n  /* Description */\n  --va-timeline-item-description-margin-top: 0.25rem;\n  --va-timeline-item-description-text-align: center;\n}\n.va-timeline-item {\n  display: var(--va-timeline-item-display);\n  flex-direction: var(--va-timeline-item-flex-direction);\n}\n.va-timeline-item__before, .va-timeline-item__after {\n  flex: 1;\n}\n.va-timeline-item--vertical {\n  flex-direction: var(--va-timeline-item-vertical-flex-direction);\n  flex-wrap: var(--va-timeline-item-vertical-flex-wrap);\n  align-items: var(--va-timeline-item-vertical-align-items);\n}\n.va-timeline-item--vertical .va-timeline-item__before,\n.va-timeline-item--vertical .va-timeline-item__after {\n  padding-top: 1rem;\n  padding-bottom: 1rem;\n}\n.va-timeline-item--vertical .va-timeline-item__before {\n  padding-right: 1rem;\n}\n.va-timeline-item--vertical .va-timeline-item__after {\n  padding-left: 1rem;\n}\n.va-timeline-item:not(.va-timeline-item--vertical) .va-timeline-item__before,\n.va-timeline-item:not(.va-timeline-item--vertical) .va-timeline-item__after {\n  padding-right: 1rem;\n  padding-left: 1rem;\n}\n.va-timeline-item:not(.va-timeline-item--vertical) .va-timeline-item__before {\n  padding-bottom: 0.5rem;\n}\n.va-timeline-item:not(.va-timeline-item--vertical) .va-timeline-item__after {\n  padding-top: 0.5rem;\n}";
const COMPONENT_NAME = "va-timeline-item";
const VaTimelineSeparatorProps = extractComponentProps(VaTimelineSeparator);
const _sfc_main = defineComponent({
  name: COMPONENT_NAME,
  props: {
    ...useComponentPresetProp,
    ...VaTimelineSeparatorProps,
    color: { type: String, default: "primary" },
    isFirst: { type: Boolean },
    isLast: { type: Boolean },
    inverted: { type: Boolean }
  },
  setup(props, { slots }) {
    const children = [
      h(
        VaTimelineSeparator,
        { ...filterComponentProps(VaTimelineSeparatorProps).value }
      )
    ];
    const before = props.inverted ? slots.after : slots.before;
    if (before) {
      children.unshift(
        h(
          "div",
          { class: `${COMPONENT_NAME}__before` },
          before()
        )
      );
    }
    const after = props.inverted ? slots.before : slots.after;
    if (after) {
      children.push(
        h(
          "div",
          { class: `${COMPONENT_NAME}__after` },
          after()
        )
      );
    }
    return () => h(
      "div",
      {
        class: [
          { [COMPONENT_NAME]: true },
          { [`${COMPONENT_NAME}--vertical`]: props.vertical },
          { [`${COMPONENT_NAME}--is-first`]: props.isFirst },
          { [`${COMPONENT_NAME}--is-last`]: props.isLast }
        ]
      },
      children
    );
  }
});
var _VaTimelineItem = /* @__PURE__ */ _export_sfc(_sfc_main, [["styles", [_style_0]]]);
export { _VaTimelineItem as _ };
