// vuestic-ui/dist/web-components/src/components/va-skeleton/VaSkeleton.js
import { defineComponent, ref, onMounted, onBeforeUnmount, computed, useCssVars, openBlock, createBlock, resolveDynamicComponent, normalizeClass, withCtx, renderSlot, createElementBlock, createCommentVNode } from "vue";
import { u as useBem } from "../../composables/useBem.js";
import { u as useColors } from "../../composables/useColors.js";
import { _ as _export_sfc } from "../../../plugin-vue_export-helper.js";
var _style_0 = ':root,\n:host {\n  --va-skeleton-wave-color: rgb(100, 100, 100);\n  --va-skeleton-wave-opacity: 0.2;\n  --va-skeleton-border-radius: unset;\n  --va-skeleton-animation-duration: 2.5s;\n}\n@keyframes pulse {\n0% {\n    filter: brightness(1);\n}\n50% {\n    filter: brightness(1);\n}\n75% {\n    filter: brightness(0.9);\n}\n100% {\n    filter: brightness(1);\n}\n}\n@keyframes wave {\n0% {\n    left: -100%;\n}\n100% {\n    left: 100%;\n}\n}\n.va-skeleton {\n  background: var(--68642958);\n  height: var(--5e0c6c8c);\n  width: var(--301414bb);\n  border-radius: var(--010d0dd4);\n  cursor: wait;\n}\n.va-skeleton--pulse {\n  animation: var(--va-skeleton-animation-duration) ease-in-out 0s infinite normal none running pulse;\n}\n.va-skeleton--wave {\n  -webkit-mask-image: -webkit-radial-gradient(white, black);\n          mask-image: -webkit-radial-gradient(white, black);\n}\n.va-skeleton--wave .va-skeleton__wave {\n  position: absolute;\n  overflow: hidden;\n  left: 0;\n  height: 100%;\n  width: 100%;\n}\n.va-skeleton--wave .va-skeleton__wave::after {\n  display: block;\n  position: relative;\n  content: "";\n  left: 0;\n  top: 0;\n  width: 100%;\n  height: 100%;\n  opacity: var(--va-skeleton-wave-opacity, 0.5);\n  background: linear-gradient(90deg, transparent, var(--va-skeleton-wave-color), transparent);\n  animation: var(--va-skeleton-animation-duration) linear 0s infinite normal none running wave;\n}\n.va-skeleton--hidden {\n  position: absolute;\n  overflow: hidden;\n  clip: rect(0 0 0 0);\n  height: 1px;\n  width: 1px;\n  margin: -1px;\n  padding: 0;\n  border: 0;\n}\n.va-skeleton--lines {\n  background: repeating-linear-gradient(180deg, var(--68642958), var(--68642958) calc(1em - var(--08c1120a) / 2), transparent calc(1em - var(--08c1120a) / 2), transparent 1em);\n}\n.va-skeleton--lines .va-skeleton__wave::after {\n  -webkit-mask-image: repeating-linear-gradient(180deg, black, black calc(1em - var(--08c1120a) / 2), transparent calc(1em - var(--08c1120a) / 2), transparent 1em);\n          mask-image: repeating-linear-gradient(180deg, black, black calc(1em - var(--08c1120a) / 2), transparent calc(1em - var(--08c1120a) / 2), transparent 1em);\n}\n.va-skeleton--text {\n  width: 100%;\n  -webkit-clip-path: polygon(0% 0%, 0% 100%, var(--5c49a604) 100%, var(--5c49a604) calc(100% - 1em), 100% calc(100% - 1em), 100% 100%, 1em 100%, var(--5c49a604) 100%, 100% 100%, 100% 0%);\n          clip-path: polygon(0% 0%, 0% 100%, var(--5c49a604) 100%, var(--5c49a604) calc(100% - 1em), 100% calc(100% - 1em), 100% 100%, 1em 100%, var(--5c49a604) 100%, 100% 100%, 100% 0%);\n}\n.va-skeleton--circle {\n  min-width: var(--301414bb);\n  min-height: var(--5e0c6c8c);\n}';
const __default__ = defineComponent({
  name: "VaSkeleton",
  props: {
    color: { type: String, default: "backgroundElement" },
    delay: { type: Number, default: 100 },
    tag: { type: String, default: "div" },
    animation: { type: String, default: "pulse" },
    lines: { type: Number, default: 1 },
    height: { type: [String], default: "5em" },
    width: { type: [String], default: "100%" },
    lineGap: { type: String, default: "8px" },
    lastLineWidth: { type: [String], default: "75%" },
    variant: { type: String, default: "squared" }
  },
  setup(props, { attrs }) {
    const doShow = ref(false);
    let timeoutId;
    onMounted(() => {
      setTimeout(() => {
        doShow.value = true;
      }, props.delay);
    });
    onBeforeUnmount(() => {
      clearTimeout(timeoutId);
    });
    const heightComputed = computed(() => {
      if (props.variant === "text") {
        return `${props.lines}em`;
      }
      return props.height;
    });
    const widthComputed = computed(() => {
      if (props.variant === "circle") {
        return heightComputed.value;
      }
      return props.width;
    });
    const { getColor } = useColors();
    const colorComputed = computed(() => getColor(props.color));
    const negativeLineGap = computed(() => `-${props.lineGap}`);
    const bem = useBem("va-skeleton", () => ({
      lines: props.lines > 1,
      text: props.variant === "text",
      circle: props.variant === "circle",
      hidden: !doShow.value,
      pulse: props.animation === "pulse",
      wave: props.animation === "wave"
    }));
    const borderRadius = computed(() => {
      if (props.variant === "circle") {
        return "50%";
      }
      if (props.variant === "rounded") {
        return `var(--va-skeleton-border-radius, calc(${heightComputed.value} / 5))`;
      }
      return "0px";
    });
    return {
      classes: computed(() => [
        ...Object.keys(bem),
        attrs.class
      ]),
      colorComputed,
      negativeLineGap,
      doShow,
      heightComputed,
      widthComputed,
      borderRadius
    };
  }
});
const __injectCSSVars__ = () => {
  useCssVars((_ctx) => ({
    "68642958": _ctx.colorComputed,
    "5e0c6c8c": _ctx.heightComputed,
    "301414bb": _ctx.widthComputed,
    "010d0dd4": _ctx.borderRadius,
    "08c1120a": _ctx.lineGap,
    "5c49a604": _ctx.lastLineWidth
  }));
};
const __setup__ = __default__.setup;
__default__.setup = __setup__ ? (props, ctx) => {
  __injectCSSVars__();
  return __setup__(props, ctx);
} : __injectCSSVars__;
const _sfc_main = __default__;
const _hoisted_1 = {
  key: 0,
  class: "va-skeleton__wave"
};
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createBlock(resolveDynamicComponent(_ctx.tag), {
    class: normalizeClass(["va-skeleton", _ctx.classes]),
    role: "status",
    "aria-live": "polite",
    "aria-label": "Loading",
    "aria-atomic": "true"
  }, {
    default: withCtx(() => [
      renderSlot(_ctx.$slots, "default"),
      _ctx.animation === "wave" ? (openBlock(), createElementBlock("div", _hoisted_1)) : createCommentVNode("", true)
    ]),
    _: 3
  }, 8, ["class"]);
}
var _VaSkeleton = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render], ["styles", [_style_0]]]);
export { _VaSkeleton as _ };
