// vuestic-ui/dist/es/src/components/va-card/components/va-card-title/VaCardTitle.js
import { defineComponent, computed, openBlock, createElementBlock, normalizeStyle, renderSlot } from "vue";
import { u as useColors } from "../../../../composables/useColors.js";
import { u as useComponentPresetProp } from "../../../../composables/useComponentPreset.js";
import { _ as _export_sfc } from "../../../../../plugin-vue_export-helper.js";
var VaCardTitle_vue_vue_type_style_index_0_scoped_true_lang = "";
const _sfc_main = defineComponent({
  name: "VaCardTitle",
  props: {
    ...useComponentPresetProp,
    textColor: { type: String }
  },
  setup(props) {
    const { getColor } = useColors();
    return {
      titleStyles: computed(() => ({
        color: props.textColor ? getColor(props.textColor) : ""
      }))
    };
  }
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("div", {
    class: "va-card-title va-card__title",
    style: normalizeStyle(_ctx.titleStyles)
  }, [
    renderSlot(_ctx.$slots, "default", {}, void 0, true)
  ], 4);
}
var VaCardTitleBase = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render], ["__scopeId", "data-v-875a3baa"]]);
export { VaCardTitleBase as V };

import './VaCardTitle.css';
