// vuestic-ui/dist/esm-node/src/services/global-config/global-config.mjs
import { j as cloneDeep_1 } from "../../../vendor.mjs";
import { ref, getCurrentInstance } from "vue";
import { g as getComponentsDefaultConfig } from "../component-config/config/default.mjs";
import { g as getIconDefaultConfig } from "../icon/config/default.mjs";
import { g as getColorDefaultConfig } from "../color/config/default.mjs";
import { g as getI18nConfigDefaults } from "../i18n/config/default.mjs";
import { g as getBreakpointDefaultConfig } from "../breakpoint/index.mjs";
import { i as inject, g as getCurrentApp } from "../current-app.mjs";
import { m as mergeDeep } from "../../utils/merge-deep.mjs";
import { g as getColorsClassesDefaultConfig } from "../colors-classes/plugin/index.mjs";
const GLOBAL_CONFIG = Symbol("GLOBAL_CONFIG");
const createGlobalConfig = () => {
  const globalConfig = ref({
    colors: getColorDefaultConfig(),
    icons: getIconDefaultConfig(),
    components: getComponentsDefaultConfig(),
    breakpoint: getBreakpointDefaultConfig(),
    i18n: getI18nConfigDefaults(),
    colorsClasses: getColorsClassesDefaultConfig()
  });
  const getGlobalConfig = () => globalConfig.value;
  const setGlobalConfig = (updater) => {
    const config = typeof updater === "function" ? updater(globalConfig.value) : updater;
    globalConfig.value = cloneDeep_1(config);
  };
  const mergeGlobalConfig = (updater) => {
    const config = typeof updater === "function" ? updater(globalConfig.value) : updater;
    globalConfig.value = mergeDeep(cloneDeep_1(globalConfig.value), config);
  };
  return {
    getGlobalConfig,
    setGlobalConfig,
    mergeGlobalConfig,
    globalConfig
  };
};
const provideForCurrentApp = (provide) => {
  var _a, _b;
  const provides = ((_a = getCurrentInstance()) == null ? void 0 : _a.appContext.provides) || ((_b = getCurrentApp()) == null ? void 0 : _b._context.provides);
  if (!provides) {
    throw new Error("Vue app not found for provide");
  }
  provides[GLOBAL_CONFIG] = provide;
  return provide;
};
function useGlobalConfig() {
  let injected = inject(GLOBAL_CONFIG);
  if (!injected) {
    injected = createGlobalConfig();
    provideForCurrentApp(injected);
  }
  return injected;
}
export { GLOBAL_CONFIG as G, createGlobalConfig as c, useGlobalConfig as u };
