// vuestic-ui/dist/es/src/components/va-button-group/VaButtonGroup.js
import { defineComponent, computed, resolveComponent, openBlock, createElementBlock, normalizeClass, createVNode, withCtx, renderSlot } from "vue";
import { e as extractComponentProps } from "../../utils/component-options/extract-component-options.js";
import { o as omit_1 } from "../../../vendor.js";
import { u as useBem } from "../../composables/useBem.js";
import { u as useComponentPresetProp } from "../../composables/useComponentPreset.js";
import { V as VaConfig } from "../va-config/VaConfig.js";
import { V as VaButton } from "../va-button/index.js";
import { _ as _export_sfc } from "../../../plugin-vue_export-helper.js";
var VaButtonGroup_vue_vue_type_style_index_0_lang = "";
const VaButtonProps = omit_1(extractComponentProps(VaButton), ["block"]);
const _sfc_main = defineComponent({
  name: "VaButtonGroup",
  components: { VaConfig },
  props: {
    ...VaButtonProps,
    ...useComponentPresetProp,
    grow: {
      type: Boolean,
      default: false
    }
  },
  setup: (props) => {
    const buttonConfig = computed(() => ({ VaButton: { ...props } }));
    const computedClass = useBem("va-button-group", () => ({
      square: !props.round,
      grow: props.grow,
      small: props.size === "small",
      large: props.size === "large"
    }));
    return { buttonConfig, computedClass };
  }
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_va_config = resolveComponent("va-config");
  return openBlock(), createElementBlock("div", {
    class: normalizeClass(["va-button-group", _ctx.computedClass])
  }, [
    createVNode(_component_va_config, { components: _ctx.buttonConfig }, {
      default: withCtx(() => [
        renderSlot(_ctx.$slots, "default")
      ]),
      _: 3
    }, 8, ["components"])
  ], 2);
}
var _VaButtonGroup = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { _VaButtonGroup as _ };

import './VaButtonGroup.css';
