// vuestic-ui/dist/web-components/src/components/va-card/components/va-card-title/VaCardTitle.js
import { defineComponent, computed, openBlock, createElementBlock, normalizeStyle, renderSlot } from "vue";
import { u as useColors } from "../../../../composables/useColors.js";
import { u as useComponentPresetProp } from "../../../../composables/useComponentPreset.js";
import { _ as _export_sfc } from "../../../../../plugin-vue_export-helper.js";
var _style_0 = ".va-card__title[data-v-875a3baa] {\n  display: flex;\n  align-items: center;\n  font-size: 0.625rem;\n  letter-spacing: 0.6px;\n  line-height: 1.2;\n  font-weight: 700;\n  text-transform: uppercase;\n}\n.va-card__title[data-v-875a3baa], .va-card__content[data-v-875a3baa], .va-card__actions[data-v-875a3baa], .va-card__actions--vertical[data-v-875a3baa] {\n  padding: var(--va-card-padding);\n}\n.va-card__title + .va-card__title[data-v-875a3baa],\n.va-card__title + .va-card__content[data-v-875a3baa],\n.va-card__title + .va-card__actions[data-v-875a3baa],\n.va-card__title + .va-card_actions__vertical[data-v-875a3baa], .va-card__content + .va-card__title[data-v-875a3baa],\n.va-card__content + .va-card__content[data-v-875a3baa],\n.va-card__content + .va-card__actions[data-v-875a3baa],\n.va-card__content + .va-card_actions__vertical[data-v-875a3baa], .va-card__actions + .va-card__title[data-v-875a3baa],\n.va-card__actions + .va-card__content[data-v-875a3baa],\n.va-card__actions + .va-card__actions[data-v-875a3baa],\n.va-card__actions + .va-card_actions__vertical[data-v-875a3baa], .va-card__actions--vertical + .va-card__title[data-v-875a3baa],\n.va-card__actions--vertical + .va-card__content[data-v-875a3baa],\n.va-card__actions--vertical + .va-card__actions[data-v-875a3baa],\n.va-card__actions--vertical + .va-card_actions__vertical[data-v-875a3baa] {\n  padding-top: 0;\n}";
const _sfc_main = defineComponent({
  name: "VaCardTitle",
  props: {
    ...useComponentPresetProp,
    textColor: { type: String }
  },
  setup(props) {
    const { getColor } = useColors();
    return {
      titleStyles: computed(() => ({
        color: props.textColor ? getColor(props.textColor) : ""
      }))
    };
  }
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("div", {
    class: "va-card-title va-card__title",
    style: normalizeStyle(_ctx.titleStyles)
  }, [
    renderSlot(_ctx.$slots, "default", {}, void 0, true)
  ], 4);
}
var VaCardTitleBase = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render], ["styles", [_style_0]], ["__scopeId", "data-v-875a3baa"]]);
export { VaCardTitleBase as V };
