// vuestic-ui/dist/es/src/components/va-inner-loading/VaInnerLoading.js
import { u as useComponentPresetProp } from "../../composables/useComponentPreset.js";
import { defineComponent, computed, resolveComponent, openBlock, createElementBlock, renderSlot, createVNode, createCommentVNode } from "vue";
import { u as useColors } from "../../composables/useColors.js";
import { u as useLoadingProps } from "../../composables/useLoading.js";
import { V as VaIcon } from "../va-icon/index.js";
import { _ as _export_sfc } from "../../../plugin-vue_export-helper.js";
var VaInnerLoading_vue_vue_type_style_index_0_lang = "";
const _sfc_main = defineComponent({
  name: "VaInnerLoading",
  components: { VaIcon },
  props: {
    ...useLoadingProps,
    ...useComponentPresetProp,
    color: { type: String },
    icon: { type: String, default: "autorenew" },
    size: { type: Number, default: 30 }
  },
  setup(props) {
    const { getColor } = useColors();
    const colorComputed = computed(() => getColor(props.color));
    return { colorComputed };
  }
});
const _hoisted_1 = { class: "inner-loading" };
const _hoisted_2 = {
  key: 0,
  class: "inner-loading__overlay",
  "aria-hidden": "true"
};
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_va_icon = resolveComponent("va-icon");
  return openBlock(), createElementBlock("div", _hoisted_1, [
    renderSlot(_ctx.$slots, "default"),
    _ctx.$props.loading ? (openBlock(), createElementBlock("div", _hoisted_2, [
      createVNode(_component_va_icon, {
        class: "inner-loading__spinner",
        spin: "",
        color: _ctx.colorComputed,
        size: _ctx.$props.size,
        name: _ctx.$props.icon
      }, null, 8, ["color", "size", "name"])
    ])) : createCommentVNode("", true)
  ]);
}
var _VaInnerLoading = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { _VaInnerLoading as _ };

import './VaInnerLoading.css';
