// vuestic-ui/dist/esm-node/src/components/va-card/components/va-card-actions/VaCardActions.mjs
import { defineComponent, openBlock, createElementBlock, normalizeClass, normalizeStyle, renderSlot } from "vue";
import { u as useAlignProps, a as useAlign } from "../../../../composables/useAlign.mjs";
import { u as useBem } from "../../../../composables/useBem.mjs";
import { p as pick_1 } from "../../../../../vendor.mjs";
import { u as useComponentPresetProp } from "../../../../composables/useComponentPreset.mjs";
import { _ as _export_sfc } from "../../../../../plugin-vue_export-helper.mjs";
var VaCardActions_vue_vue_type_style_index_0_lang = "";
const _sfc_main = defineComponent({
  name: "VaCardActions",
  props: {
    ...useAlignProps,
    ...useComponentPresetProp
  },
  setup(props) {
    const { alignComputed } = useAlign(props);
    const classComputed = useBem("va-card__actions", () => ({
      ...pick_1(props, ["vertical"])
    }));
    return {
      classComputed,
      alignComputed
    };
  }
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("div", {
    class: normalizeClass(["va-card__actions", _ctx.classComputed]),
    style: normalizeStyle(_ctx.alignComputed)
  }, [
    renderSlot(_ctx.$slots, "default")
  ], 6);
}
var VaCardActionsBase = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { VaCardActionsBase as V };
