// vuestic-ui/dist/web-components/src/components/va-skeleton/components/VaSkeletonGroup.js
import { defineComponent, ref, onMounted, onBeforeMount, computed, resolveComponent, openBlock, createBlock, withCtx, createElementVNode, mergeProps, renderSlot } from "vue";
import { u as useBem } from "../../../composables/useBem.js";
import { V as VaConfig } from "../../va-config/VaConfig.js";
import { _ as _export_sfc } from "../../../../plugin-vue_export-helper.js";
var _style_0 = ".va-skeleton-group[data-v-91b5ab0a] {\n  position: relative;\n  cursor: progress;\n}\n.va-skeleton-group--hidden[data-v-91b5ab0a] {\n  display: none;\n}";
const _sfc_main = defineComponent({
  name: "VaSkeletonGroup",
  components: {
    VaConfig
  },
  props: {
    color: { type: String, default: "backgroundElement" },
    delay: { type: Number, default: 100 },
    animation: { type: String, default: "pulse" },
    lines: { type: Number, default: 1 },
    lineGap: { type: String, default: "8px" },
    lastLineWidth: { type: [String], default: "75%" }
  },
  setup(props) {
    const doShow = ref(false);
    let timeoutId;
    onMounted(() => {
      timeoutId = setTimeout(() => {
        doShow.value = true;
      }, props.delay);
    });
    onBeforeMount(() => {
      clearTimeout(timeoutId);
    });
    const bem = useBem("va-skeleton-group", () => ({
      hidden: doShow.value === false
    }));
    return {
      doShow,
      bem,
      config: computed(() => ({ ...props, delay: 0 }))
    };
  }
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_VaConfig = resolveComponent("VaConfig");
  return openBlock(), createBlock(_component_VaConfig, {
    components: { VaSkeleton: _ctx.config }
  }, {
    default: withCtx(() => [
      createElementVNode("div", mergeProps({
        class: ["va-skeleton-group", _ctx.bem]
      }, _ctx.$attrs), [
        renderSlot(_ctx.$slots, "default", {}, void 0, true)
      ], 16)
    ]),
    _: 3
  }, 8, ["components"]);
}
var _VaSkeletonGroup = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render], ["styles", [_style_0]], ["__scopeId", "data-v-91b5ab0a"]]);
export { _VaSkeletonGroup as _ };
