// vuestic-ui/dist/web-components/src/components/va-card/components/va-card-actions/VaCardActions.js
import { defineComponent, openBlock, createElementBlock, normalizeClass, normalizeStyle, renderSlot } from "vue";
import { u as useAlignProps, a as useAlign } from "../../../../composables/useAlign.js";
import { u as useBem } from "../../../../composables/useBem.js";
import { p as pick_1 } from "../../../../../vendor.js";
import { u as useComponentPresetProp } from "../../../../composables/useComponentPreset.js";
import { _ as _export_sfc } from "../../../../../plugin-vue_export-helper.js";
var _style_0 = ".va-card__title, .va-card__content, .va-card__actions, .va-card__actions--vertical {\n  padding: var(--va-card-padding);\n}\n.va-card__title + .va-card__title,\n.va-card__title + .va-card__content,\n.va-card__title + .va-card__actions,\n.va-card__title + .va-card_actions__vertical, .va-card__content + .va-card__title,\n.va-card__content + .va-card__content,\n.va-card__content + .va-card__actions,\n.va-card__content + .va-card_actions__vertical, .va-card__actions + .va-card__title,\n.va-card__actions + .va-card__content,\n.va-card__actions + .va-card__actions,\n.va-card__actions + .va-card_actions__vertical, .va-card__actions--vertical + .va-card__title,\n.va-card__actions--vertical + .va-card__content,\n.va-card__actions--vertical + .va-card__actions,\n.va-card__actions--vertical + .va-card_actions__vertical {\n  padding-top: 0;\n}\n.va-card__actions button,\n.va-card__actions .va-button {\n  margin: 0 var(--va-card-actions-btn-margin);\n}\n.va-card__actions button:first-child,\n.va-card__actions .va-button:first-child {\n  margin-left: 0;\n}\n.va-card__actions button:last-child,\n.va-card__actions .va-button:last-child {\n  margin-right: 0;\n}\n.va-card__actions--vertical button,\n.va-card__actions--vertical .va-button {\n  margin: var(--va-card-actions-btn-margin) 0;\n}\n.va-card__actions--vertical button:first-child,\n.va-card__actions--vertical .va-button:first-child {\n  margin-top: 0;\n}\n.va-card__actions--vertical button:last-child,\n.va-card__actions--vertical .va-button:last-child {\n  margin-bottom: 0;\n}";
const _sfc_main = defineComponent({
  name: "VaCardActions",
  props: {
    ...useAlignProps,
    ...useComponentPresetProp
  },
  setup(props) {
    const { alignComputed } = useAlign(props);
    const classComputed = useBem("va-card__actions", () => ({
      ...pick_1(props, ["vertical"])
    }));
    return {
      classComputed,
      alignComputed
    };
  }
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("div", {
    class: normalizeClass(["va-card__actions", _ctx.classComputed]),
    style: normalizeStyle(_ctx.alignComputed)
  }, [
    renderSlot(_ctx.$slots, "default")
  ], 6);
}
var VaCardActionsBase = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render], ["styles", [_style_0]]]);
export { VaCardActionsBase as V };
