// vuestic-ui/dist/web-components/src/components/va-list/VaList.js
import { defineComponent, computed, openBlock, createElementBlock, normalizeClass, renderSlot } from "vue";
import { u as useComponentPresetProp } from "../../composables/useComponentPreset.js";
import { _ as _export_sfc } from "../../../plugin-vue_export-helper.js";
var _style_0 = ":root,\n:host {\n  --va-list-width: 100%;\n  /* Label */\n  --va-list-label-text-align: center;\n  --va-list-label-padding: 0.3rem 0;\n  /* Item */\n  --va-list-item-display: flex;\n  --va-list-item-align-items: center;\n  --va-list-item-width: 100%;\n  --va-list-item-height: 100%;\n  /* Item label */\n  --va-list-item-label-color: #34495e;\n  --va-list-item-label-display: -webkit-box;\n  --va-list-item-label-box-orient: vertical;\n  --va-list-item-label-overflow: hidden;\n  --va-list-item-label-line-height: normal;\n  --va-list-item-label-caption-font-size: 85%;\n  --va-list-item-label-caption-color: #babfc2;\n  /* Item Section */\n  --va-list-item-section-display: flex;\n  --va-list-item-section-icon-min-width: 1.5rem;\n  --va-list-item-section-icon-align-items: center;\n  --va-list-item-section-icon-justify-content: center;\n  --va-list-item-section-icon-margin: 0.6rem 0.75rem;\n  --va-list-item-section-icon-font-size: 1.25rem;\n  --va-list-item-section-avatar-min-width: 3rem;\n}\n.va-list {\n  width: var(--va-list-width);\n  font-family: var(--va-font-family);\n}\n.va-list--fit {\n  width: -moz-fit-content;\n  width: fit-content;\n}";
const _sfc_main = defineComponent({
  name: "VaList",
  props: {
    ...useComponentPresetProp,
    fit: { type: Boolean, default: false }
  },
  setup: (props) => ({
    computedClass: computed(() => ({ "va-list--fit": props.fit }))
  })
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("div", {
    class: normalizeClass(["va-list", _ctx.computedClass]),
    role: "list"
  }, [
    renderSlot(_ctx.$slots, "default")
  ], 2);
}
var _VaList = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render], ["styles", [_style_0]]]);
export { _VaList as _ };
