// vuestic-ui/dist/web-components/src/components/va-list/VaListSeparator.js
import { defineComponent, computed, openBlock, createElementBlock, normalizeClass } from "vue";
import { u as useComponentPresetProp } from "../../composables/useComponentPreset.js";
import { _ as _export_sfc } from "../../../plugin-vue_export-helper.js";
var _style_0 = ".va-list-separator {\n  height: 1px;\n  background-color: #e6e9ec;\n}\n.va-list-separator--offset {\n  margin-left: 1.5rem;\n}\n.va-list-separator--spaced {\n  margin: 1.5rem 0;\n}";
const _sfc_main = defineComponent({
  name: "VaListSeparator",
  props: {
    ...useComponentPresetProp,
    fit: { type: Boolean, default: false },
    spaced: { type: Boolean, default: false }
  },
  setup: (props) => ({
    computedClass: computed(() => ({
      "va-list-separator--offset": !props.fit,
      "va-list-separator--spaced": props.spaced
    }))
  })
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("div", {
    "aria-hidden": "true",
    class: normalizeClass(["va-list-separator", _ctx.computedClass])
  }, null, 2);
}
var _VaListSeparator = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render], ["styles", [_style_0]]]);
export { _VaListSeparator as _ };
