// vuestic-ui/dist/web-components/src/components/va-modal/VaModal.js
import { w as withConfigTransport } from "../../services/config-transport/withConfigTransport.js";
import { defineComponent, h, Transition, shallowRef, toRef, computed, watchEffect, watch, onMounted, nextTick, resolveComponent, openBlock, createElementBlock, normalizeClass, renderSlot, normalizeProps, guardReactiveProps, createCommentVNode, createBlock, Teleport, createVNode, mergeProps, withCtx, normalizeStyle, createElementVNode, withKeys, Fragment, toDisplayString, createTextVNode } from "vue";
import { u as useClickOutside } from "../../composables/useClickOutside.js";
import { u as useColors } from "../../composables/useColors.js";
import { u as useComponentPresetProp } from "../../composables/useComponentPreset.js";
import { u as useDocument } from "../../composables/useDocument.js";
import { u as useWindow } from "../../composables/useWindow.js";
import { u as useModalLevel } from "../../composables/useModalLevel.js";
import { u as useStatefulEmits, a as useStatefulProps, b as useStateful } from "../../composables/useStateful.js";
import { u as useTextColor } from "../../composables/useTextColor.js";
import { u as useTranslation } from "../../composables/useTranslation.js";
import { u as useTrapFocus } from "../../composables/useTrapFocus.js";
import { V as VaButton } from "../va-button/index.js";
import { V as VaIcon } from "../va-icon/index.js";
import { u as useBlur } from "./hooks/useBlur.js";
import { _ as _export_sfc } from "../../../plugin-vue_export-helper.js";
var _style_0 = ":root,\n:host {\n  --va-modal-title-margin-bottom: 1.5rem;\n  --va-modal-position: fixed;\n  --va-modal-top: 0;\n  --va-modal-left: 0;\n  --va-modal-display: flex;\n  --va-modal-width: 100%;\n  --va-modal-height: 100%;\n  --va-modal-align-items: center;\n  --va-modal-justify-content: center;\n  --va-modal-overflow: hidden;\n  --va-modal-outline: 0;\n  --va-modal-z-index: var(--va-z-index-teleport-overlay);\n  --va-modal-container-z-index: 100;\n  --va-modal-basic-duration: 0.3s;\n  --va-modal-opacity-transition: calc(var(--va-modal-basic-duration) * 0.5) cubic-bezier(1, 0.5, 0.8, 1);\n  --va-modal-transform-transition: var(--va-modal-basic-duration) ease;\n  --va-modal-overlay-background-blur-radius: 4px;\n  --va-modal-padding: var(--va-modal-padding-top) var(--va-modal-padding-right) var(--va-modal-padding-bottom) var(--va-modal-padding-left);\n  --va-modal-padding-top: 1.25rem;\n  --va-modal-padding-right: 1.5rem;\n  --va-modal-padding-bottom: 1.5rem;\n  --va-modal-padding-left: 1.5rem;\n  /* Dialog */\n  --va-modal-dialog-min-height: 3.125rem;\n  --va-modal-dialog-height: fit-content;\n  --va-modal-dialog-border-radius: 0.375rem;\n  --va-modal-dialog-margin: 1rem;\n  --va-modal-dialog-box-shadow: 0 4px 70px -18px var(--va-shadow);\n  --va-modal-dialog-max-width: 768px;\n  --va-modal-dialog-max-height: calc(100vh - 2rem);\n  --va-modal-dialog-position: relative;\n  /* Fullscreen */\n  --va-modal-overlay-position: fixed;\n  --va-modal-overlay-top: 0;\n  --va-modal-overlay-left: 0;\n  --va-modal-overlay-z-index: calc(var(--va-modal-container-z-index) - 1);\n  --va-modal-overlay-width: 100vw;\n  --va-modal-overlay-height: 100vh;\n  --va-modal-overlay-opacity-transition: opacity math.div(2 * var(--va-modal-basic-duration), 3) cubic-bezier(1, 0.5, 0.8, 1);\n}\n.va-modal-overlay-background--blurred > :not(div[class*=va-]) {\n  filter: blur(var(--va-modal-overlay-background-blur-radius));\n  position: absolute;\n  height: 100%;\n  width: 100%;\n}\n.va-modal {\n  position: var(--va-modal-position);\n  display: var(--va-modal-display);\n  align-items: var(--va-modal-align-items);\n  justify-content: var(--va-modal-justify-content);\n  width: var(--va-modal-width);\n  height: var(--va-modal-height);\n  top: var(--va-modal-top);\n  left: var(--va-modal-left);\n  overflow: var(--va-modal-overflow);\n  outline: var(--va-modal-outline);\n  z-index: var(--va-modal-z-index);\n  font-family: var(--va-font-family);\n}\n.va-modal__title {\n  margin-bottom: 1.5rem;\n  font-size: 0.625rem;\n  letter-spacing: 0.6px;\n  line-height: 1.2;\n  font-weight: 700;\n  text-transform: uppercase;\n}\n.va-modal__container {\n  z-index: var(--va-modal-container-z-index);\n  max-width: 100%;\n}\n.va-modal-enter-from .va-modal__container, .va-modal-leave-to .va-modal__container {\n  opacity: 0;\n  transform: translateY(-30%);\n}\n.va-modal-enter-active .va-modal__container, .va-modal-leave-active .va-modal__container {\n  transition: opacity var(--va-modal-opacity-transition), transform var(--va-modal-transform-transition);\n}\n.va-modal__dialog {\n  min-height: var(--va-modal-dialog-min-height);\n  height: var(--va-modal-dialog-height);\n  border-radius: var(--va-modal-dialog-border-radius, var(--va-block-border-radius));\n  margin: var(--va-modal-dialog-margin);\n  box-shadow: var(--va-modal-dialog-box-shadow, var(--va-block-box-shadow));\n  max-width: var(--va-modal-dialog-max-width);\n  max-height: var(--va-modal-dialog-max-height);\n  position: var(--va-modal-dialog-position);\n  overflow: auto;\n}\n.va-modal__overlay {\n  position: var(--va-modal-overlay-position);\n  top: var(--va-modal-overlay-top);\n  left: var(--va-modal-overlay-left);\n  z-index: var(--va-modal-overlay-z-index);\n  width: var(--va-modal-overlay-width);\n  height: var(--va-modal-overlay-height);\n}\n.va-modal-enter-from .va-modal__overlay, .va-modal-leave-to .va-modal__overlay {\n  opacity: 0;\n}\n.va-modal-enter-active .va-modal__overlay, .va-modal-leave-active .va-modal_overlay {\n  transition: var(--va-modal-overlay-opacity-transition);\n}\n.va-modal--fullscreen {\n  min-width: 100vw !important;\n  min-height: 100vh !important;\n  border-radius: 0;\n  margin: 0;\n}\n@media all and (max-width: 576px) {\n.va-modal--mobile-fullscreen {\n    margin: 0 !important;\n    min-width: 100vw !important;\n    min-height: 100vh !important;\n    border-radius: 0;\n}\n}\n.va-modal--size-small {\n  max-width: 576px;\n}\n@media all and (max-width: 576px) {\n.va-modal--size-small {\n    max-width: 100vw !important;\n}\n}\n.va-modal--size-small .va-modal__inner {\n  max-width: 576px;\n}\n@media all and (max-width: 576px) {\n.va-modal--size-small .va-modal__inner {\n    max-width: 100vw !important;\n}\n}\n.va-modal--size-large {\n  max-width: 992px;\n}\n.va-modal--size-large .va-modal__inner {\n  max-width: 992px;\n}\n.va-modal--fixed-layout .va-modal__inner {\n  overflow: hidden;\n  padding: var(--va-modal-padding-top) 0 var(--va-modal-padding-bottom);\n  max-height: calc(100vh - 2rem);\n}\n.va-modal--fixed-layout .va-modal__inner .va-modal__header,\n.va-modal--fixed-layout .va-modal__inner .va-modal__footer,\n.va-modal--fixed-layout .va-modal__inner .va-modal__title {\n  padding: 0 var(--va-modal-padding-right) 0 var(--va-modal-padding-left);\n}\n.va-modal--fixed-layout .va-modal__inner .va-modal__message {\n  padding: 0 var(--va-modal-padding-right) 0 var(--va-modal-padding-left);\n  overflow: auto;\n}\n.va-modal--fixed-layout .va-modal__dialog {\n  overflow: hidden;\n}\n.va-modal--no-padding .va-modal__inner {\n  padding: 0;\n}\n.va-modal__message {\n  margin-bottom: 1.5rem;\n}\n.va-modal__inner {\n  overflow: visible;\n  display: flex;\n  position: relative;\n  flex-flow: column;\n  padding: var(--va-modal-padding);\n  max-width: 768px;\n  margin: auto;\n}\n.va-modal__inner > div:last-of-type {\n  margin-bottom: 0;\n}\n.va-modal__close {\n  position: absolute;\n  top: 1rem;\n  right: 1rem;\n  cursor: pointer;\n  font-size: 1.5rem;\n  font-style: normal;\n  color: var(--va-secondary);\n  z-index: 1;\n}\n.va-modal__default-cancel-button {\n  margin-right: 0.75rem;\n}\n.va-modal__footer {\n  margin-top: auto;\n  min-height: -moz-fit-content;\n  min-height: fit-content;\n  display: flex;\n  flex-wrap: wrap;\n  justify-content: center;\n}\n.va-modal__footer:last-of-type {\n  margin-bottom: 0;\n}";
const ModalElement = defineComponent({
  name: "ModalElement",
  inheritAttrs: false,
  props: {
    ...useComponentPresetProp,
    isTransition: { type: Boolean, default: true }
  },
  setup: (props, { slots, attrs }) => () => {
    var _a;
    return props.isTransition ? h(Transition, { ...attrs }, slots) : (_a = slots.default) == null ? void 0 : _a.call(slots, attrs);
  }
});
const _sfc_main = defineComponent({
  name: "VaModal",
  inheritAttrs: false,
  components: { VaButton, VaIcon, ModalElement },
  emits: [
    ...useStatefulEmits,
    "cancel",
    "ok",
    "before-open",
    "open",
    "before-close",
    "close",
    "click-outside"
  ],
  props: {
    ...useStatefulProps,
    modelValue: { type: Boolean, default: false },
    attachElement: { type: String, default: "body" },
    allowBodyScroll: { type: Boolean, default: false },
    disableAttachment: { type: Boolean, default: false },
    title: { type: String, default: "" },
    message: { type: String, default: "" },
    okText: { type: String, default: "$t:ok" },
    cancelText: { type: String, default: "$t:cancel" },
    hideDefaultActions: { type: Boolean, default: false },
    fullscreen: { type: Boolean, default: false },
    mobileFullscreen: { type: Boolean, default: true },
    noDismiss: { type: Boolean, default: false },
    noOutsideDismiss: { type: Boolean, default: false },
    noEscDismiss: { type: Boolean, default: false },
    maxWidth: { type: String, default: "" },
    maxHeight: { type: String, default: "" },
    anchorClass: { type: String },
    size: {
      type: String,
      default: "medium",
      validator: (value) => ["medium", "small", "large"].includes(value)
    },
    fixedLayout: { type: Boolean, default: false },
    withoutTransitions: { type: Boolean, default: false },
    overlay: { type: Boolean, default: true },
    overlayOpacity: { type: [Number, String], default: 0.6 },
    blur: { type: Boolean, default: false },
    zIndex: { type: [Number, String], default: void 0 },
    backgroundColor: { type: String, default: "background-secondary" },
    noPadding: { type: Boolean, default: false },
    beforeClose: { type: Function }
  },
  setup(props, { emit }) {
    const rootElement = shallowRef();
    const modalDialog = shallowRef();
    const { trapFocusIn, freeFocus } = useTrapFocus();
    const {
      registerModal,
      unregisterModal,
      isTopLevelModal,
      isLowestLevelModal
    } = useModalLevel();
    const { getColor } = useColors();
    const { textColorComputed } = useTextColor(toRef(props, "backgroundColor"));
    const { valueComputed } = useStateful(props, emit);
    const computedClass = computed(() => ({
      "va-modal--fullscreen": props.fullscreen,
      "va-modal--mobile-fullscreen": props.mobileFullscreen,
      "va-modal--fixed-layout": props.fixedLayout,
      "va-modal--no-padding": props.noPadding,
      [`va-modal--size-${props.size}`]: props.size !== "medium"
    }));
    const computedModalContainerStyle = computed(() => ({ "z-index": props.zIndex }));
    const computedDialogStyle = computed(() => ({
      maxWidth: props.maxWidth,
      maxHeight: props.maxHeight,
      color: textColorComputed.value,
      background: getColor(props.backgroundColor)
    }));
    const computedOverlayStyles = computed(() => {
      if (!props.overlay || !isLowestLevelModal.value) {
        return;
      }
      return {
        "background-color": `rgba(0, 0, 0, ${props.overlayOpacity})`,
        "z-index": props.zIndex && Number(props.zIndex) - 1
      };
    });
    const show = () => {
      valueComputed.value = true;
    };
    const hide = (cb) => {
      const _hide = () => {
        valueComputed.value = false;
        cb == null ? void 0 : cb();
      };
      props.beforeClose ? props.beforeClose(_hide) : _hide();
    };
    const toggle = () => {
      valueComputed.value = !valueComputed.value;
    };
    const cancel = () => {
      hide(() => emit("cancel"));
    };
    const ok = () => {
      hide(() => emit("ok"));
    };
    const trapFocusInModal = () => {
      nextTick(() => {
        if (modalDialog.value) {
          trapFocusIn(modalDialog.value);
        }
      });
    };
    const onBeforeEnterTransition = (el) => emit("before-open", el);
    const onAfterEnterTransition = (el) => emit("open", el);
    const onBeforeLeaveTransition = (el) => emit("before-close", el);
    const onAfterLeaveTransition = (el) => emit("close", el);
    const listenKeyUp = (e) => {
      const hideModal = () => {
        if (e.code === "Escape" && !props.noEscDismiss && !props.noDismiss && isTopLevelModal.value) {
          cancel();
        }
      };
      setTimeout(hideModal);
    };
    useClickOutside([modalDialog], () => {
      if (!valueComputed.value || props.noOutsideDismiss || props.noDismiss || !isTopLevelModal.value) {
        return;
      }
      emit("click-outside");
      cancel();
    });
    const window = useWindow();
    watchEffect(() => {
      var _a, _b;
      if (valueComputed.value) {
        (_a = window.value) == null ? void 0 : _a.addEventListener("keyup", listenKeyUp);
      } else {
        (_b = window.value) == null ? void 0 : _b.removeEventListener("keyup", listenKeyUp);
      }
    });
    useBlur(toRef(props, "blur"), valueComputed);
    const documentRef = useDocument();
    const setBodyOverflow = (overflow) => {
      if (!documentRef.value || props.allowBodyScroll) {
        return;
      }
      documentRef.value.body.style.overflow = overflow;
    };
    watch(valueComputed, (newValueComputed) => {
      if (newValueComputed) {
        registerModal();
        setBodyOverflow("scroll hidden");
        return;
      }
      if (isLowestLevelModal.value) {
        freeFocus();
        setBodyOverflow("");
      }
      unregisterModal();
    });
    watch(isTopLevelModal, (newIsTopLevelModal) => {
      if (newIsTopLevelModal) {
        trapFocusInModal();
      }
    });
    onMounted(() => {
      if (valueComputed.value) {
        registerModal();
      }
    });
    const publicMethods = {
      ...useTranslation(),
      show,
      hide,
      toggle,
      cancel,
      ok,
      onBeforeEnterTransition,
      onAfterEnterTransition,
      onBeforeLeaveTransition,
      onAfterLeaveTransition,
      listenKeyUp
    };
    return {
      getColor,
      rootElement,
      modalDialog,
      valueComputed,
      computedClass,
      computedDialogStyle,
      computedModalContainerStyle,
      computedOverlayStyles,
      ...publicMethods
    };
  }
});
const _hoisted_1 = ["aria-labelledby"];
const _hoisted_2 = {
  key: 0,
  class: "va-modal__anchor"
};
const _hoisted_3 = {
  key: 0,
  class: "va-modal"
};
const _hoisted_4 = { key: 0 };
const _hoisted_5 = {
  key: 1,
  class: "va-modal__header"
};
const _hoisted_6 = {
  key: 2,
  class: "va-modal__message"
};
const _hoisted_7 = {
  key: 3,
  class: "va-modal__message"
};
const _hoisted_8 = {
  key: 4,
  class: "va-modal__footer"
};
const _hoisted_9 = {
  key: 5,
  class: "va-modal__footer"
};
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_va_icon = resolveComponent("va-icon");
  const _component_va_button = resolveComponent("va-button");
  const _component_modal_element = resolveComponent("modal-element");
  return openBlock(), createElementBlock("div", {
    ref: "rootElement",
    class: normalizeClass(["va-modal-entry", _ctx.$props.anchorClass]),
    role: "dialog",
    "aria-modal": "true",
    "aria-labelledby": _ctx.title
  }, [
    _ctx.$slots.anchor ? (openBlock(), createElementBlock("div", _hoisted_2, [
      renderSlot(_ctx.$slots, "anchor", normalizeProps(guardReactiveProps({ show: _ctx.show, hide: _ctx.hide, toggle: _ctx.toggle })))
    ])) : createCommentVNode("", true),
    (openBlock(), createBlock(Teleport, {
      to: _ctx.attachElement,
      disabled: _ctx.$props.disableAttachment
    }, [
      createVNode(_component_modal_element, mergeProps({
        name: "va-modal",
        isTransition: !_ctx.$props.withoutTransitions,
        appear: "",
        duration: 300
      }, _ctx.$attrs, {
        onBeforeEnter: _ctx.onBeforeEnterTransition,
        onAfterEnter: _ctx.onAfterEnterTransition,
        onBeforeLeave: _ctx.onBeforeLeaveTransition,
        onAfterLeave: _ctx.onAfterLeaveTransition
      }), {
        default: withCtx(() => [
          _ctx.valueComputed ? (openBlock(), createElementBlock("div", _hoisted_3, [
            _ctx.$props.overlay ? (openBlock(), createElementBlock("div", {
              key: 0,
              class: "va-modal__overlay",
              style: normalizeStyle(_ctx.computedOverlayStyles)
            }, null, 4)) : createCommentVNode("", true),
            createElementVNode("div", {
              class: "va-modal__container",
              style: normalizeStyle(_ctx.computedModalContainerStyle)
            }, [
              createElementVNode("div", {
                ref: "modalDialog",
                class: normalizeClass(["va-modal__dialog", _ctx.computedClass]),
                style: normalizeStyle(_ctx.computedDialogStyle)
              }, [
                _ctx.$props.fullscreen ? (openBlock(), createBlock(_component_va_icon, {
                  key: 0,
                  name: "va-close",
                  class: "va-modal__close",
                  role: "button",
                  "aria-label": _ctx.t("close"),
                  tabindex: "0",
                  onClick: _ctx.cancel,
                  onKeydown: [
                    withKeys(_ctx.cancel, ["space"]),
                    withKeys(_ctx.cancel, ["enter"])
                  ]
                }, null, 8, ["aria-label", "onClick", "onKeydown"])) : createCommentVNode("", true),
                createElementVNode("div", {
                  class: "va-modal__inner",
                  style: normalizeStyle({ maxWidth: _ctx.$props.maxWidth, maxHeight: _ctx.$props.maxHeight })
                }, [
                  _ctx.$slots.content ? (openBlock(), createElementBlock("div", _hoisted_4, [
                    renderSlot(_ctx.$slots, "content", normalizeProps(guardReactiveProps({ cancel: _ctx.cancel, ok: _ctx.ok })))
                  ])) : createCommentVNode("", true),
                  !_ctx.$slots.content ? (openBlock(), createElementBlock(Fragment, { key: 1 }, [
                    _ctx.title ? (openBlock(), createElementBlock("div", {
                      key: 0,
                      class: "va-modal__title",
                      style: normalizeStyle({ color: _ctx.getColor("primary") })
                    }, toDisplayString(_ctx.$props.title), 5)) : createCommentVNode("", true),
                    _ctx.$slots.header ? (openBlock(), createElementBlock("div", _hoisted_5, [
                      renderSlot(_ctx.$slots, "header")
                    ])) : createCommentVNode("", true),
                    _ctx.$props.message ? (openBlock(), createElementBlock("div", _hoisted_6, toDisplayString(_ctx.$props.message), 1)) : createCommentVNode("", true),
                    _ctx.$slots.default ? (openBlock(), createElementBlock("div", _hoisted_7, [
                      renderSlot(_ctx.$slots, "default")
                    ])) : createCommentVNode("", true),
                    (_ctx.$props.cancelText || _ctx.$props.okText) && !_ctx.$props.hideDefaultActions ? (openBlock(), createElementBlock("div", _hoisted_8, [
                      _ctx.$props.cancelText ? (openBlock(), createBlock(_component_va_button, {
                        key: 0,
                        preset: "secondary",
                        color: "secondary",
                        class: "va-modal__default-cancel-button",
                        onClick: _ctx.cancel
                      }, {
                        default: withCtx(() => [
                          createTextVNode(toDisplayString(_ctx.tp(_ctx.$props.cancelText)), 1)
                        ]),
                        _: 1
                      }, 8, ["onClick"])) : createCommentVNode("", true),
                      createVNode(_component_va_button, { onClick: _ctx.ok }, {
                        default: withCtx(() => [
                          createTextVNode(toDisplayString(_ctx.tp(_ctx.$props.okText)), 1)
                        ]),
                        _: 1
                      }, 8, ["onClick"])
                    ])) : createCommentVNode("", true),
                    _ctx.$slots.footer ? (openBlock(), createElementBlock("div", _hoisted_9, [
                      renderSlot(_ctx.$slots, "footer")
                    ])) : createCommentVNode("", true)
                  ], 64)) : createCommentVNode("", true)
                ], 4)
              ], 6)
            ], 4)
          ])) : createCommentVNode("", true)
        ]),
        _: 3
      }, 16, ["isTransition", "onBeforeEnter", "onAfterEnter", "onBeforeLeave", "onAfterLeave"])
    ], 8, ["to", "disabled"]))
  ], 10, _hoisted_1);
}
var _VaModal = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render], ["styles", [...(   VaButton.styles || []), ...( VaIcon.styles || []), ...( ModalElement .styles || []), _style_0]]]);
const VaModal = withConfigTransport(_VaModal);
export { VaModal as V };
