// vuestic-ui/dist/es/src/components/va-file-upload/VaFileUploadUndo/VaFileUploadUndo.js
import { defineComponent, ref, computed, onMounted, useCssVars, resolveComponent, openBlock, createElementBlock, Fragment, createVNode, createElementVNode, normalizeClass, toDisplayString, withCtx, createTextVNode } from "vue";
import { u as useBem } from "../../../composables/useBem.js";
import { u as useStrictInject } from "../../../composables/useStrictInject.js";
import { V as VaFileUploadKey } from "../types.js";
import { V as VaButton } from "../../va-button/index.js";
import { V as VaProgressBar } from "../../va-progress-bar/index.js";
import { _ as _export_sfc } from "../../../../plugin-vue_export-helper.js";
var VaFileUploadUndo_vue_vue_type_style_index_0_lang = "";
const INJECTION_ERROR_MESSAGE = "The VaFileUploadUndo component should be used in the context of VaFileUpload component";
const __default__ = defineComponent({
  name: "VaFileUploadUndo",
  components: {
    VaProgressBar,
    VaButton
  },
  props: {
    vertical: { type: Boolean, default: false }
  },
  emits: ["recover"],
  setup: (props) => {
    const progress = ref(100);
    const {
      undoDuration,
      undoButtonText,
      deletedFileMessage
    } = useStrictInject(VaFileUploadKey, INJECTION_ERROR_MESSAGE);
    const computedClasses = useBem("va-file-upload-undo", () => ({
      vertical: props.vertical
    }));
    const undoDurationStyle = computed(() => {
      var _a;
      return `${(_a = undoDuration.value) != null ? _a : 0}ms`;
    });
    onMounted(() => {
      const timer = setTimeout(() => {
        progress.value = 0;
        clearTimeout(timer);
      }, 0);
    });
    return {
      progress,
      undoDuration,
      undoButtonText,
      computedClasses,
      undoDurationStyle,
      deletedFileMessage
    };
  }
});
const __injectCSSVars__ = () => {
  useCssVars((_ctx) => ({
    "38c05944": _ctx.undoDurationStyle
  }));
};
const __setup__ = __default__.setup;
__default__.setup = __setup__ ? (props, ctx) => {
  __injectCSSVars__();
  return __setup__(props, ctx);
} : __injectCSSVars__;
const _sfc_main = __default__;
const _hoisted_1 = { class: "va-file-upload-undo__text" };
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_va_progress_bar = resolveComponent("va-progress-bar");
  const _component_va_button = resolveComponent("va-button");
  return openBlock(), createElementBlock(Fragment, null, [
    createVNode(_component_va_progress_bar, {
      "model-value": _ctx.progress,
      rounded: false,
      class: "va-file-upload-undo-progress-bar"
    }, null, 8, ["model-value"]),
    createElementVNode("div", {
      class: normalizeClass(["va-file-upload-undo", _ctx.computedClasses])
    }, [
      createElementVNode("span", _hoisted_1, toDisplayString(_ctx.deletedFileMessage), 1),
      createVNode(_component_va_button, {
        class: "va-file-upload-undo__button",
        "aria-label": _ctx.undoButtonText,
        size: "small",
        outline: "",
        onClick: _cache[0] || (_cache[0] = ($event) => _ctx.$emit("recover"))
      }, {
        default: withCtx(() => [
          createTextVNode(toDisplayString(_ctx.undoButtonText), 1)
        ]),
        _: 1
      }, 8, ["aria-label"])
    ], 2)
  ], 64);
}
var _VaFileUploadUndo = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { _VaFileUploadUndo as _ };

import './VaFileUploadUndo.css';
