// vuestic-ui/dist/es/src/components/va-rating/components/VaRatingItemNumberButton.js
import { defineComponent, openBlock, createElementBlock, normalizeStyle, toDisplayString } from "vue";
import { u as useSizeProps, a as useSize } from "../../../composables/useSize.js";
import { u as useVaRatingColorsProps, a as useVaRatingColors } from "../hooks/useVaRatingColors.js";
import { _ as _export_sfc } from "../../../../plugin-vue_export-helper.js";
const _sfc_main = defineComponent({
  name: "VaRatingItemNumberButton",
  props: {
    ...useVaRatingColorsProps,
    ...useSizeProps,
    itemNumber: { type: Number, required: true },
    modelValue: { type: Number, required: true }
  },
  setup(props) {
    return {
      ...useVaRatingColors(props),
      ...useSize(props, "VaRating")
    };
  }
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("button", {
    class: "va-rating__number-item",
    tabindex: "-1",
    "aria-hidden": "true",
    style: normalizeStyle({
      background: _ctx.backgroundComputed,
      color: _ctx.textColorComputed,
      width: _ctx.sizeComputed,
      height: _ctx.sizeComputed,
      fontSize: _ctx.fontSizeComputed,
      borderRadius: `${parseInt(_ctx.fontSizeComputed) * 0.125}rem`
    })
  }, toDisplayString(_ctx.itemNumber), 5);
}
var VaRatingItemNumberButton = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { VaRatingItemNumberButton as V };
