// vuestic-ui/dist/web-components/src/components/va-card/VaCard.js
import { defineComponent, computed, useCssVars, openBlock, createBlock, resolveDynamicComponent, normalizeClass, normalizeStyle, withCtx, renderSlot } from "vue";
import { p as pick_1 } from "../../../vendor.js";
import { g as getGradientBackground } from "../../services/color/utils.js";
import { u as useBem } from "../../composables/useBem.js";
import { u as useColors } from "../../composables/useColors.js";
import { u as useComponentPresetProp } from "../../composables/useComponentPreset.js";
import { u as useRouterLinkProps, a as useRouterLink } from "../../composables/useRouterLink.js";
import { u as useTextColor } from "../../composables/useTextColor.js";
import { _ as _export_sfc } from "../../../plugin-vue_export-helper.js";
var _style_0 = ':root,\n:host {\n  --va-card-display: block;\n  --va-card-position: relative;\n  --va-card-overflow: visible;\n  --va-card-box-shadow: var(--va-box-shadow);\n  --va-card-border-radius: 0.375rem;\n  --va-card-color: #34495e;\n  --va-card-background-color: var(--va-background-secondary);\n  --va-card-padding: 1.25rem;\n  /* Outlined */\n  --va-card-outlined-border: 3px solid var(--va-background-element);\n  --va-card-outlined-box-shadow: none;\n  /* Stripe */\n  --va-card-stripe-border-size: var(--va-stripe-border-size);\n  /* Dark */\n  --va-card-dark-color: #ffffff;\n  --va-card-dark-background-color: #34495e;\n  /* Actions */\n  --va-card-actions-btn-margin: 4px;\n}\n.va-card {\n  display: var(--va-card-display);\n  position: var(--va-card-position);\n  overflow: var(--va-card-overflow);\n  box-shadow: var(--va-card-box-shadow, var(--va-block-box-shadow));\n  border-radius: var(--va-card-border-radius, var(--va-block-border-radius));\n  color: var(--va-card-color);\n  background-color: var(--va-card-background-color);\n  font-family: var(--va-font-family);\n}\n.va-card > div:first-child {\n  border-top-right-radius: var(--va-card-border-radius);\n  border-top-left-radius: var(--va-card-border-radius);\n}\n.va-card > div:last-child {\n  border-bottom-right-radius: var(--va-card-border-radius);\n  border-bottom-left-radius: var(--va-card-border-radius);\n}\n.va-card--square {\n  border-radius: 0;\n}\n.va-card--outlined {\n  box-shadow: var(--va-card-outlined-box-shadow);\n  border: var(--va-card-outlined-border, var(--va-block-border));\n}\n.va-card--no-border {\n  border: none;\n}\n.va-card--disabled {\n  cursor: default;\n  opacity: 0.4;\n  pointer-events: none;\n  -webkit-user-select: none;\n     -moz-user-select: none;\n          user-select: none;\n}\n.va-card--link {\n  cursor: pointer;\n}\n.va-card--stripe::after {\n  content: "";\n  position: absolute;\n  width: 100%;\n  height: var(--va-card-stripe-border-size);\n  top: 0;\n  left: 0;\n  background: var(--0e93fe42);\n  border-top-right-radius: var(--va-card-border-radius);\n  border-top-left-radius: var(--va-card-border-radius);\n}';
const __default__ = defineComponent({
  name: "VaCard",
  props: {
    ...useRouterLinkProps,
    ...useComponentPresetProp,
    tag: { type: String, default: "div" },
    square: { type: Boolean, default: false },
    outlined: { type: Boolean, default: false },
    bordered: { type: Boolean, default: true },
    disabled: { type: Boolean, default: false },
    href: { type: String, default: "" },
    target: { type: String, default: "" },
    stripe: { type: Boolean, default: false },
    stripeColor: { type: String, default: "" },
    gradient: { type: Boolean, default: false },
    textColor: { type: String },
    color: { type: String, default: "background-secondary" }
  },
  setup(props) {
    const { getColor } = useColors();
    const { isLinkTag, tagComputed, hrefComputed } = useRouterLink(props);
    const { textColorComputed } = useTextColor();
    const stripeColorComputed = computed(() => getColor(props.stripeColor));
    const classComputed = useBem("va-card", () => ({
      ...pick_1(props, ["square", "outlined", "disabled", "stripe"]),
      noBorder: !props.bordered,
      link: isLinkTag.value
    }));
    const cardStyles = computed(() => {
      const background = props.gradient && props.color ? getGradientBackground(getColor(props.color)) : getColor(props.color);
      return {
        background,
        color: textColorComputed.value
      };
    });
    return {
      classComputed,
      cardStyles,
      stripeColorComputed,
      tagComputed,
      hrefComputed
    };
  }
});
const __injectCSSVars__ = () => {
  useCssVars((_ctx) => ({
    "0e93fe42": _ctx.stripeColorComputed
  }));
};
const __setup__ = __default__.setup;
__default__.setup = __setup__ ? (props, ctx) => {
  __injectCSSVars__();
  return __setup__(props, ctx);
} : __injectCSSVars__;
const _sfc_main = __default__;
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createBlock(resolveDynamicComponent(_ctx.tagComputed), {
    class: normalizeClass(["va-card", _ctx.classComputed]),
    style: normalizeStyle(_ctx.cardStyles),
    href: _ctx.hrefComputed,
    target: _ctx.target,
    to: _ctx.to,
    replace: _ctx.replace,
    exact: _ctx.exact,
    "active-class": _ctx.activeClass,
    "exact-active-class": _ctx.exactActiveClass
  }, {
    default: withCtx(() => [
      renderSlot(_ctx.$slots, "default")
    ]),
    _: 3
  }, 8, ["class", "style", "href", "target", "to", "replace", "exact", "active-class", "exact-active-class"]);
}
var _VaCard = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render], ["styles", [_style_0]]]);
export { _VaCard as _ };
