// vuestic-ui/dist/web-components/src/components/va-timeline/VaTimelineSeparator/VaTimelineSeparator.js
import { defineComponent, h } from "vue";
import { u as useColors } from "../../../composables/useColors.js";
import { u as useComponentPresetProp } from "../../../composables/useComponentPreset.js";
import { _ as _export_sfc } from "../../../../plugin-vue_export-helper.js";
var _style_0 = ":root,\n:host {\n  --va-timeline-separator-display: flex;\n  --va-timeline-separator-align-items: center;\n  /* Line */\n  --va-timeline-separator-line-transition: background-color ease 0.5s;\n  --va-timeline-separator-line-width: 0.125rem;\n  --va-timeline-separator-line-height: 0.125rem;\n  --va-timeline-separator-line-flex: 1;\n  /* Center */\n  --va-timeline-separator-center-transition: background-color ease 0.5s;\n  --va-timeline-separator-center-width: 0.5rem;\n  --va-timeline-separator-center-height: 0.5rem;\n  --va-timeline-separator-center-border-radius: 0.5rem;\n  /* Vertical */\n  --va-timeline-separator-vertical-flex-direction: column;\n}\n.va-timeline-separator {\n  display: var(--va-timeline-separator-display);\n  align-items: var(--va-timeline-separator-align-items);\n}\n.va-timeline-separator--vertical {\n  flex-direction: var(--va-timeline-separator-vertical-flex-direction);\n}\n.va-timeline-separator__line {\n  transition: var(--va-timeline-separator-line-transition);\n  width: var(--va-timeline-separator-line-width);\n  height: var(--va-timeline-separator-line-height);\n  flex: var(--va-timeline-separator-line-flex);\n}\n.va-timeline-separator__center {\n  transition: var(--va-timeline-separator-center-transition);\n  width: var(--va-timeline-separator-center-width);\n  height: var(--va-timeline-separator-center-height);\n  border-radius: var(--va-timeline-separator-center-border-radius);\n}";
const componentName = "va-timeline-separator";
const _sfc_main = defineComponent({
  name: componentName,
  props: {
    ...useComponentPresetProp,
    color: { type: String, default: "primary" },
    vertical: { type: Boolean },
    active: { type: Boolean },
    activePrevious: { type: Boolean },
    activeNext: { type: Boolean }
  },
  setup(props) {
    const { getColor } = useColors();
    return () => h(
      "div",
      {
        class: {
          [componentName]: true,
          [`${componentName}--vertical`]: props.vertical
        }
      },
      [
        h("div", {
          class: {
            [`${componentName}__line`]: true,
            [`${componentName}__line--active`]: props.activePrevious
          },
          style: {
            backgroundColor: getColor(props.activePrevious ? props.color : "divider")
          }
        }),
        h("div", {
          class: {
            [`${componentName}__center`]: true,
            [`${componentName}__center--active`]: props.active
          },
          style: {
            backgroundColor: getColor(props.active ? props.color : "divider")
          }
        }),
        h("div", {
          class: {
            [`${componentName}__line`]: true,
            [`${componentName}__line--active`]: props.activeNext
          },
          style: {
            backgroundColor: getColor(props.activeNext ? props.color : "divider")
          }
        })
      ]
    );
  }
});
var _VaTimelineSeparator = /* @__PURE__ */ _export_sfc(_sfc_main, [["styles", [_style_0]]]);
export { _VaTimelineSeparator as _ };
